maven/lib/dependabot/maven/version_parser.rb

```diff
@@ -0,0 +1,137 @@
+        parse_version(false)
+
+        raise ArgumentError, "Malformed version string #{version}" if parse_result.to_a.empty?
+
+        T.must(parse_result)
+      end
+
+      private
+
+      sig { returns(StringScanner) }
+      attr_reader :scanner
+
+      sig { returns(String) }
+      attr_reader :version
+
+      sig { returns(T.nilable(TokenBucket)) }
+      attr_reader :parse_result
+
+      sig { params(token: T.nilable(T.any(String, Integer))).void }
+      def parse_addition(token = nil)
+        @token_bucket&.addition = TokenBucket.new(tokens: [token].compact)
+        @token_bucket = @token_bucket&.addition
+
+        scanner.skip(/-+/)
+        parse_version(true)
+      end
+
+      sig { params(number_begins_partition: T.nilable(T::Boolean)).void }
+      def parse_version(number_begins_partition) # rubocop:disable Metrics/AbcSize,Metrics/CyclomaticComplexity,Metrics/MethodLength,Metrics/PerceivedComplexity
+        # skip leading v if any
+        scanner.skip(/v/)
+
+        until scanner.eos?
+          if (s = scanner.scan(/\d+/))
+            if number_begins_partition
+              parse_addition(s.to_i)
+            else
+              T.must(@token_bucket).tokens << s.to_i
+            end
+
+          elsif (s = scanner.match?(/a\d+/))
+            # aN is equivalent to alpha-N
+            scanner.skip("a")
+            parse_addition(ALPHA)
+
+          elsif (s = scanner.match?(/b\d+/))
+            # bN is equivalent to beta-N
+            scanner.skip("b")
+            parse_addition(BETA)
+
+          elsif (s = scanner.match?(/m\d+/))
+            # mN is equivalent to milestone-N
+            scanner.skip("m")
+            parse_addition(MILESTONE)
+
+          elsif (s = scanner.scan(/(alpha|beta|milestone|rc|cr|sp|ga|final|release|snapshot)[a-z]+/))
+            # process "alpha" and others as normal lexical tokens if they're followed by a letter
+            parse_addition(s)
+
+          elsif (s = scanner.scan("alpha"))
+            # handle alphaN, alpha-X, alpha.X, or ending alpha
+            parse_addition(ALPHA)
+
+          elsif (s = scanner.scan("beta"))
+            parse_addition(BETA)
+          elsif (s = scanner.scan("milestone"))
+            parse_addition(MILESTONE)
+
+          elsif (s = scanner.scan(/(rc|cr)/))
+            parse_addition(RC)
+
+          elsif (s = scanner.scan("snapshot"))
+            parse_addition(SNAPSHOT)
+
+          elsif (s = scanner.scan(/ga|final|release/))
+            parse_addition
+
+          elsif (s = scanner.scan("sp"))
+            parse_addition(SP)
+
+          # `+` is parsed as an addition as stated in maven version spec
+          elsif (s = scanner.scan(/[a-z_+]+/))
+            parse_addition(s)
+
+          elsif (s = scanner.scan("."))
+            number_begins_partition = false
+
+          elsif (s = scanner.scan("-"))
+            number_begins_partition = true
+
+          else
+            raise ArgumentError, scanner.rest
```

Comment:
should this be `MalformedVersionError` or something like that? Tagging @sachin-sandhu

Reply:
@amazimbe , @abdulapopoola , closest one i could find is `BadRequirementError` for illformed versions

Reply:
@abdulapopoola @sachin-sandhu As can be seen here: https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb#L223 , the original gem version uses ArgumentError. BadRequirementError is just a subclass of argument error customised for requirement.rb ( https://github.com/rubygems/rubygems/blob/master/lib/rubygems/requirement.rb#L46 ). So what I did here is probably more inline with the original intentions.

Reply:
Yes, we should use BadRequirementError since it is already piped through our systems and will show up as a handled error.

Reply:
@amazimbe , we try to use errors in **errors.rb** , that way the (handled) errors are presented to user and they don't end up in Sentry as _unknow_error_